### ISSUE TITLE
large job launch does not work [well

### ISSUE THREAD

**Issue by trws** [flux-framework/flux-core#1406]

**Comment by grondo** [comment]
For this test you might try `-o stdio-delay-commit`, though 400 lines of output shouldn't take that long anyway.

Does `-O output-file` work for smaller jobs? Is the real issue here "redirecting output of large jobs doesn't work"?

**Comment by trws** [comment]
It does work for small jobs, I can try delay commit, but the issues are 
that it didn’t get output and also that it took 16 minutes.  
Especially given that all of the brokers are already up, and doing it 
with tools that don’t have pre-existing connections takes 6 seconds, 
we should really figure out why it takes so long.

**Comment by trws** [comment]
Just realized I had forgotten to post this: It is an 800-node instance.  Running the same by using mpiexec set to rsh to all the target nodes simultaneously from a single source node, takes 6 seconds for the same list of nodes.

**Comment by grondo** [comment]
Yeah, unfortunately launching with state in the kvs is going to be slower than rsh direct no matter what we do. I'm worried that a lot of extra kvs commits may have slipped into launch of high rank-count jobs, since we haven't had a chance to try that in a long time. (last I checked it was nowhere near this bad though)

**Comment by grondo** [comment]
Might be useful to also grab the `ljw.` dmesg lines from all nodes. It is surprising that it took 4m to get to the "running" state (barrier/fence after all wrexecds have finished starting all tasks).

**Comment by trws** [comment]
I'd be happy to do that if I knew how... What I posted was all that came out of running dmesg, is there a way to grab the rest of it? The instance is still up.  Also, I tried once with both delay commit and no-pmi turned on and the time was almost identical.  Oddly the master broker is almost completely idle the whole time too, there's something funny going on.

**Comment by grondo** [comment]
Oh, @trws, sorry to be so chatty, but can you try `flux-wreckrun --immediate` with the same configuration, and verify that `rank.N.cores` information makes sense for this job? I just tried haha test oversubscribing on ipa with 400 brokers and `flux wreckrun -v -N 400 -n400 -w complete` ran in <2s
```
$ flux wreckrun -v -N 400 -n 400 -w complete hostname
wreckrun: 0.012s: Registered jobid 4
wreckrun: 0.014s: State = reserved
wreckrun: 0.015s: job.submit: Function not implemented
wreckrun: Allocating 400 tasks across 400 available nodes..
wreckrun: tasks per node: node[0-399]: 1
wreckrun: 0.085s: Sending run event
wreckrun: 1.463s: State = starting
wreckrun: 1.559s: State = running
wreckrun: 1.559s: State = complete
```

Of course, I think brokers on shared nodes use a different topology that might make this test a lot faster but @garlick might have to comment on that.

**Comment by trws** [comment]
I think I can do that, this version still has the using wreckrun kills sched bug, but I think I can just reload sched.

**Comment by grondo** [comment]
I agree there must be something strange going on.  Sorry about this!

**Comment by trws** [comment]
Ok, something seriously strange is going on, now completely for sure:

```
splash:hwloc$ flux wreckrun -v -N 400 -n 400 -w complete hostname
wreckrun: 0.013s: Registered jobid 5
wreckrun: 0.014s: State = reserved
wreckrun: Allocating 400 tasks across 884 available nodes..
wreckrun: tasks per node: node[0-399]: 1
wreckrun: 0.150s: Sending run event
wreckrun: 1.523s: State = starting
wreckrun: 3.142s: State = running
wreckrun: 3.142s: State = complete
```

**Comment by trws** [comment]
I'm trying the same wreckrun with IO allowed, will send the output when (if..?) it finishes.

**Comment by grondo** [comment]
I sadly got an assertion error from `zuuid_new()` when trying to process output from 400 tasks (memory error?) Hopefully doesn't hit you.

**Comment by trws** [comment]
Could be a memory error, but sierra nodes have 256gb each, so that shouldn't be an issue.  Will see.

**Comment by trws** [comment]
Wow... just to be completely sure, I ran the wreckrun version redirecting to files, the whole thing completes in two seconds, and all commands ran to completion.  The one set to retrieve output is still running.

**Comment by grondo** [comment]
Just to understand 

 * `flux wreckrun -I -N400 -w complete hostname` works
 * `flux wreckrun -I -N400 hostname` works
 * `flux wreckrun -I -N400 -O hostfile.output hostname` works?
 * `flux submit -N400 -O hostfile.output` runs slow

Maybe something sched sets in the `rank.N.cores` dirs is causing slowness or triggering a bug ?

**Comment by dongahn** [comment]
> Maybe something sched sets in the rank.N.cores dirs is causing slowness

Would be good to verify this. If this is the issue, the patch I posted can be worth a try.

**Comment by trws** [comment]
What does -I do?

otherwise, -w complete works
The second and below run sufficiently slowly that I haven't been able to complete any of them in the time we've been talking.

**Comment by dongahn** [comment]
Seems it is more likely IO? It could be a few commit sites in the execution service could be a scalability bottleneck. Maybe redirecting to files could be a valid work around for the weekend production runs though.

**Comment by grondo** [comment]
`-I` runs without invoking the scheduler (I'm assuming you're running off current master)

> Seems it is more likely IO? 

The same I/O is happening with and without `-w complete`, the only difference is that something isn't trying to read the I/O in realtime from the kvs. The reader in the `-O output` case could be stuck which would explain the "hang". If this is the case then just dump the I/O to a file after the job is complete.

**Comment by grondo** [comment]
i.e., I should say it appears that *reading* I/O has a bug causing the slowness, not writing of I/O nor the KVS commits.

**Comment by trws** [comment]
This is a bit behind while I test the current master to make sure nothing regressed before making it production.  

Yeah, it pretty much has to be I/O.  Or something caused by reading it maybe?  This is the output from running without -w:

```
splash:hwloc$ flux wreckrun -v -N 400 -n 400 hostname
wreckrun: 0.012s: Registered jobid 6
wreckrun: 0.013s: State = reserved
wreckrun: Allocating 400 tasks across 884 available nodes..
wreckrun: tasks per node: node[0-399]: 1
wreckrun: 0.153s: Sending run event
wreckrun: 1149.840s: State = starting
wreckrun: 1154.960s: State = running
wreckrun: 1154.981s: State = complete
```

Note that it ran for another 100 seconds after printing that it was complete.

**Comment by dongahn** [comment]
@trws: sched master doesn't have that N.cores fix. If you are still using hwloc with 1Node containing only one core, that should be okay. Otherwise yoh may want to try the N.cores fix.

**Comment by grondo** [comment]
I wonder if the output reader is being woken up a lot and eating up process time so other reactor callbacks are starved. Does `flux wreck ls` show slow time as well for this particular run?

**Comment by dongahn** [comment]
https://github.com/flux-framework/flux-core/issues/1400#issuecomment-377603026

**Comment by dongahn** [comment]
I won't be able to respond for a while. In transit to LA.

**Comment by trws** [comment]
It shows 14 minutes for everything except the ones with the `-w` argument.  Oddly that's two minutes faster than the submitted version.

**Comment by dongahn** [comment]
There is always scheduling cost with submit version...

**Comment by trws** [comment]
True, but the surprise is because sched sent the run request in less than half a second.

**Comment by grondo** [comment]
I'm baffled how use of `-w` alone would change the timing here. That just disables reading of output/err and opening of single stdin kz file in the kvs. It should have had little effect on the running job. 

For now it might be best to avoid redirecting output within a job until we figure this out.  (There are workarounds since the output is stored in the KVS). My availability is going to be spotty for the next week.

Unfortunately I can't reproduce on 400 brokers oversubscribed on 4 nodes of ipa.

**Comment by trws** [comment]
You made a very good point that helps me a lot @grondo.  Using submit without `-O` makes the performance issue completely go away.  Launch of 400 procs on 400 nodes in less than 3 seconds, really excellent!

**Comment by grondo** [comment]
Hm, sorry about the I/O issue.

Just so you have the complete picture, the way output redirection works for wreck is that the same code used for I/O in  `flux-wreckrun` is run inside the reactor on the first wrexecd of the job. If something went haywire in one or many of those (400) callbacks it perhaps is interrupting the other  event handling in that loop, and could stall the job from progressing.

We'll continue to track the output handling bug in this issue if that is Ok.

**Comment by trws** [comment]
Works for me, thanks for helping track this down.

**Comment by trws** [comment]
A bit more data here.  Trying it with 175 nodes, it takes 5 seconds, with 180 it takes minutes.  It really looks a lot like starvation.  Here's a snippet of perf report output:

```
-   73.79%     0.00%  lua      flux.so                    [.] kvswatch_cb                                             ▒
     kvswatch_cb                                                                                                      ▒
   - iowatcher_kz_ready_cb                                                                                            ▒
      - 73.65% kz_get                                                                                                 ▒
         - 73.47% getnext                                                                                             ▒
            - 68.10% flux_kvs_lookup_get                                                                              ▒
               - 68.06% decode_treeobj                                                                                ▒
                  - 68.05% flux_rpc_get_unpack                                                                        ▒
                     - 67.73% flux_future_get                                                                         ▒
                        - flux_future_wait_for                                                                        ▒
                           - 44.52% flux_dispatch_requeue                                                             ▒
                              - 44.51% dispatch_requeue.isra.5.part.6                                                 ▒
                                 + 27.41% flux_requeue                                                                ▒
                                 + 15.75% flux_msg_destroy                                                            ▒
                                 + 1.29% zlist_pop                                                                    ▒
                           + 22.92% flux_reactor_run                                                                  ▒
            + 4.80% flux_future_destroy                                                                              
```

It looks like requeues account for about half of the time, but this is in a wait on a future inside a callback, I would have expected it to be fulfilled already before the callback is invoked.  Could be I'm reading this wrong though.  Any kvs/watcher insight @garlick?

**Comment by garlick** [comment]
I will have to look at this in the morning., sorry.

**Comment by trws** [comment]
That probably explains it.  I was misreading that loop as "get all the output that's ready now without blocking, then stop."  Is there an alternate version of kz_get that would give us that behavior?

Dropping the loop would probably help. Another stopgap might be to have an option to create an output file per task and directly pipe output to it locally in wreck.  I realize that's, at best, unkind to the filesystem, but it should be relatively simple to prototype and avoid the entire kz code path.

**Comment by garlick** [comment]
It seems like the `kz_ready_f` callback _could_ be deferred until a KVS lookup continuation has fetched the new data.  Avoiding the synchronous lookup there might take care of the excessive message re-queues.

I will have a closer look at that, but I did want to get clarification where we think this is occurring.  Is it in `wrexecd.c::kz_stdin()` (wrexecd's kz_read_f callback)?

**Comment by trws** [comment]
The loop we were talking about is in the flux-lua.c file, in iowatcher_kz_ready_cb. I think the kz_stdin callback is the one doing the output connection from the task to kz, which seems ok, it's the ready data notification end that seems to be causing issues.

**Comment by grondo** [comment]
The plroblem is in the lua kz reader code. We have ntasks*2 readers in
wreckrun itself, or wrexecd on first rank of job when -O filename is used.
Sorry will try to help more in a bit

On Apr 1, 2018 9:22 AM, "Jim Garlick" <notifications@github.com> wrote:

It seems like the kz_ready_f callback *could* be deferred until a KVS
lookup continuation has fetched the new data. Avoiding the synchronous
lookup there might take care of the excessive message re-queues.

I will have a closer look at that, but I did want to get clarification
where we think this is occurring. Is it in wrexecd.c::kz_stdin() (wrexecd's
kz_read_f callback)?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/flux-framework/flux-core/issues/1406#issuecomment-377797999>,
or mute the thread

**Comment by trws** [comment]
Funny thought, but do kvs directory watch callbacks get invoked for subdirectories?  Also, did the atomically append operation ever happen? Either of those should make it possible to make it one reader regardless of num tasks. (probably a do-later item, but since we're in there now it came to mind)

**Comment by garlick** [comment]
Got it thanks!

Then it does sound like the high value fix is to change kz so that its internal KVS watch callback calls `flux_kvs_lookup()` on the new data, and then a continuation of the lookup calls the `kz_ready_f` callback when the data is available.  `kz_get()` would fetch data cached in the kz handle and should return EWOULDBLOCK when open in non-blocking mode and no more data is cached.

Sound right?

**Comment by garlick** [comment]
@trws, yes on both, though that would require the watcher to have some knowledge of how the ranks are organized in a job directory.  Something to ponder.

**Comment by grondo** [comment]
> Sound right?

Yes, anything that avoids these watchers causing sleeping in reactor callbacks. I think all the Lua callbacks expect is one chunk of output to be handed to them.

Eventually we plan to do I/O reduction in which case there probably will be a single reader for all output, which will help. As we write the new execution system, removal of all N:N patterns should be a goal. (Besides output files, we also have per-task directories which should go away in the replacement)

Sorry to leave you with this issue @garlick

**Comment by garlick** [comment]
Sorry, I'm getting confused.  Could you define "with output redirected"?  Why is that the slow case?

**Comment by grondo** [comment]
@garlick, output redirected is slow because of it attaches all kz watchers to the reactor of the first wrexecd in the job.

**Comment by trws** [comment]
When using '-O' on submit or wreckrun, or generically adding the output key to the jobdesc, it causes either the wreckrun command or the main wreck daemon to register a watch for every output and error stream.  Without that option, these commands all run quite quickly and well, despite the same amount of output going into the kvs.

This is output I got from perf record on a wreckrun of 300 processes, each with its own node:

```
      - 98.73% luaD_precall                                                                                           ▒
         - 87.91% l_iowatcher_add                                                                                     ▒
            - 87.73% kz_set_ready_cb                                                                                  ▒
               - 87.73% flux_kvs_watch_dir                                                                            ▒
                  - 86.63% kvs_watch_rpc_get                                                                          ▒
                     - 86.58% flux_rpc_get_unpack                                                                     ▒
                        - 86.46% flux_future_get                                                                      ▒
                           - flux_future_wait_for                                                                     ▒
                              + 55.82% flux_dispatch_requeue     
```

It looks like the majority of the time in wreckrun for this one was actually spent in just setting up the watches in the first place.  Do we need the synchronous get in the watch_dir function?

**Comment by garlick** [comment]
> Do we need the synchronous get in the watch_dir function?

Hmm, yes, in fact we could have it return a future like everything else.  Let me see about that.

So is the "go fast" option then `flux wreckrun --wait-until=completed`?

**Comment by trws** [comment]
Yes, or submit without '-O' also works fast.

**Comment by grondo** [comment]
> So is the "go fast" option then flux wreckrun --wait-until=completed

> Yes, or submit without '-O' also works fast.

Anything trying to read the stderr/out from a large job appears to be slow. Even `flux wreck attach ID` for a large job (though in my 4K task tests it completed much faster than @trws's 400 task example)

When testing with `flux wreck attach` there does appear to be a large startup cost. The program doesn't go back into the reactor loop until it has added watchers for every task stdout/err, so during this time messages for the kz watchers we've already added are probably coming in and getting requeued.

Anything that speeds up this initial registration and gets us into the reactor faster will help, but also keep in mind how much time we want to spend improving this

**Comment by grondo** [comment]
In case it wasn't clear from before, when using submit with `-O` the first wrexecd of the job takes over the output reading (and writing it to a file). So perhaps things are worse there because kz reading starves the wrexecd from processing other work related to job management.

**Comment by garlick** [comment]
Fixing the watch interface (at least for kz) seems like a high value, yet fairly contained effort.

Yeah, at some point it makes sense to do the I/O reduction idea and stop propping up kz.  Even if we get rid of all the synchronous RPC's that are blocking the reactor, I shudder at how much processing takes place in the KVS module for each watched key.

Thanks for the clarification guys!

**Comment by grondo** [comment]
Yeah, sorry @garlick I didn't mean to discourage or minimize your excellent efforts here. My point was more about avoiding the use of `-O file` for jobs if it is painful. I'm a bit worried that even for 400 tasks @trws's case is so slow. In my testing on ipa I was getting output from 4K tasks in ~3m not 10m for a job 10x smaller!

**Comment by trws** [comment]
I get the impression that the individual latency from having all of them on different nodes is contributing to it being this bad. Possibly also from having one broker per task.

**Comment by grondo** [comment]
> I get the impression that the individual latency from having all of them on different nodes is contributing to it being this bad. Possibly also from having one broker per task.

I might not be understanding your statement, but with `-O file` *all* kz watchers on on the first wrexecd of the job, just like in wreckrun they are all running in the wreckrun process. I did try 1 task per "broker" on ipa and couldn't reproduce the extreme results in the sierra case:

```
$ flux getattr size
400
$ time flux wreckrun -N400 hostname > /dev/null

real	0m5.498s
user	0m2.201s
sys	0m2.281s
```

**Comment by trws** [comment]
That pretty much means it has to be nodes then.  I know all the watchers are on one node, but the kvs is being hammered by all of the nodes, so anything blocking will have more messages to requeue before it finds a match.  There may be more to it than that, but there's something specific about it being individual nodes it seems.

**Comment by grondo** [comment]
@trws true, and I think brokers that share a node wire up differently (forgot that for a minute).

However, what baffles me is that if we don't install the kvs watchers everything is fine ... i.e. writing to kvs not a problem, which is the only thing done from multiple nodes. (probably missing something simple though)

**Comment by garlick** [comment]
The re-queueing only occurs in the flux_t handle/endpoint, not in the broker.  The amount of re-queueing would be proportional to the number of messages arrving at that endpoint, not passing through its broker.  (forget for a moment that modules are threads of the broker - they only communicate with the broker through messages).

If that seemed like a non-sequitur maybe I'm not understanding your point @trws.

**Comment by trws** [comment]
I see what you mean @garlick, but the more traffic there is and the higher the latency of the traffic the more the queues will fill up.  I'm expecting that all of the kvs messages, including the watch and subsequent get, go to the kvs endpoint don't they?

**Comment by garlick** [comment]
Yes, all true.  Sorry if I misunderstood your earlier description.

**Comment by garlick** [comment]
Rereading this I still am not sure we are communicating.  If you're around today @trws let's have a chat.  I want to make sure I'm working on the highest priority problem.

**Comment by grondo** [comment]
What's the status of this issue? I think things should have greatly improved since @garlick's work in libkz, however I don't have a good place to test. I could never reproduce this issue on up to 512 brokers oversubscribed over just a few nodes.

**Comment by trws** [comment]
It’s a little better, but only a little.  The fundamental problem with 
this one is that getting all the watchers setup causes some really nasty 
feedback with job launch that can make the two together take a very long 
time, and sometimes lose output among other things.  It only seems to 
happen at truly large scales though, so it’s a bit hard to test.

**Comment by grondo** [comment]
Thanks for the feedback @trws! I think we know that the 2*ntasks "streams" per job isn't going to work and have plans to fix for scale with a new I/O scheme. Hopefully the workarounds are tolerable for the splash use case and sorry about the issues!

**Comment by trws** [comment]
No worries, we’ve gotten it to where we’ve worked around this for 
splash.
